### PR TITLE
Add support for ImageTexture3D serialization

### LINKED ISSUE #80866: ImageTexture3D serialization bug
### Godot version

4.1 stable and 4.2.dev3

### System information

Godot v4.1.stable - Windows 10.0.22621 - Vulkan (Mobile) - integrated Intel(R) Iris(R) Plus Graphics (Intel Corporation; 27.20.100.9621) - Intel(R) Core(TM) i7-1065G7 CPU @ 1.30GHz (8 Threads)

### Issue description

When I create a ImageTexture3D, populate it and then save it I obtain a 1x1x1 Lum8 pink texture instead of the one I created (I think it is properly created beacause it behaves well when I use it in a custom shader).

This happens whether 
- it is running as tool or not 
- I try to save in a res or in a tres file
- I put flags in the `ResourceSaver.save` function or not

and it don't give errors.

And when It is in a shader uniform in editor, working properly - If I save the scene and reopen it, there's pink where the texture is involved and the 1x1x1 in the uniform field.

### Steps to reproduce

- Create a project and a scene
- Give the root the script below
- Close the scene and reopen it to make sure that `@tool` takes effect
- Try running the scene or checking the `save` boolean field in the inspector 
- Check if the `ImageTexture3D` got saved properly in the FileSystem Dock

```
@tool
extends Node

@export var save := false :
	set(v) :
		ResourceSaver.save(_get_grad_tex(), "res://tex3D.res")
		ResourceSaver.save(_get_grad_tex(), "res://tex3D.tres")

func _ready() -> void :
	if Engine.is_editor_hint() : return
	ResourceSaver.save(_get_grad_tex(), "res://tex3D.tres")
	ResourceSaver.save(_get_grad_tex(), "res://tex3D.res")

func _get_grad_tex() -> ImageTexture3D:
	var res := ImageTexture3D.new()
	var imgs : Array[Image] = [] 
	for i in 10 :
		var img := Image.create(10, 10, false, Image.FORMAT_RGB8)
		imgs.push_back(img)
	for x in 10 :
		for z in 10 :
			for y in 10 :
				imgs[z].set_pixel(x, y, Color(x/10., y/10., z/10.))
	res.create(Image.FORMAT_RGB8, 10, 10, 10, false, imgs )
	return res
```

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by jsjtxietian**
Fixes #80866

In the issue comment, @clayjohn posted a partially working diff to support ImageTexture3D serilization. I took it and modifed the code to make it work correctly (at least in the MRP provided).

If anyone needs this function, feel free to take it and help test this code. I would appreciate any feedback or alternative suggestions.

### CODE REVIEW

File: scene/resources/image_texture.cpp [hunk 1]
```diff
@@ -462,9 +462,54 @@ void ImageTexture3D::set_path(const String &p_path, bool p_take_over) {
...
+	Image::Format new_format = img_base->get_format();
+	int new_width = img_base->get_width();
+	int new_height = img_base->get_height();
+	int new_depth = 0;
+	bool new_mipmaps = false;
+
+	for (int i = 1; i < p_images.size(); i++) {
+		Ref<Image> img = p_images[i];
+		ERR_FAIL_COND(img.is_null());
+		ERR_FAIL_COND_MSG(img->get_format() != new_format, "All images must share the same format");
```
Comment (AThousandShips) [inline scene/resources/image_texture.cpp:490]:
```suggestion
		ERR_FAIL_COND_MSG(img->get_format() != new_format, "All images must share the same format.");
```

File: scene/resources/image_texture.cpp [hunk 2]
```diff
@@ -462,9 +462,54 @@ void ImageTexture3D::set_path(const String &p_path, bool p_take_over) {
...
+				new_depth = i;
+			}
+		}
+	}
+
+	if (new_depth == 0) {
+		new_depth = p_images.size();
+	}
+
+	ERR_FAIL_COND(_create(new_format, new_width, new_height, new_depth, new_mipmaps, p_images) != OK);
```
Comment (AThousandShips) [inline scene/resources/image_texture.cpp:504]:
```suggestion
	Error err = _create(new_format, new_width, new_height, new_depth, new_mipmaps, p_images);
	ERR_FAIL_COND(err != OK);
```
Avoid putting methods with side-effects in an error macro AFAIK, it is done above but I think we should avoid it.

Comment (jsjtxietian) [reply scene/resources/image_texture.cpp:504]: > Avoid putting methods with side-effects in an error macro AFAIK

Thanks, it makes sense.

**Comment by bitsawer** [comment]
Those errors might come when the binary or text resource loader tries to get the property default value (`ClassDB::class_get_default_property_value()`). Adding a simple test to `_get_images()` might be enough as we are lazily allocating texture RID's and returning an empty array is fine. So it would become like this:

```cpp
TypedArray<Image> ImageTexture3D::_get_images() const {
	TypedArray<Image> images;
	if (texture.is_valid()) {
		Vector<Ref<Image>> raw_images = get_data();
		ERR_FAIL_COND_V(raw_images.is_empty(), TypedArray<Image>());

		for (int i = 0; i < raw_images.size(); i++) {
			images.push_back(raw_images[i]);
		}
	}
	return images;
}
```

ImageTextureLayered might possibly need a similar check if used in a same manner as the ImageTexture3D in the issue code snippet, but that would need some more testing and is a bit out of scope of this PR. 

Kind of speculating here, so needs testing for confirmation.

**Comment by jsjtxietian** [comment]
Thanks! It solved my problem.

**Review by bitsawer — APPROVED** [review]
Tested locally on top of current master (9e455f424a635549b295272d82757c7ff99f33d7), seems to work and fixes the linked issue and serialization.

@clayjohn might still want to take a final look as it is mostly his suggested code.

**Review by clayjohn — APPROVED** [review]
Looks fine to me. This isn't really my area of expertise, but I also tested locally and it appears to work fine.

**Comment by akien-mga** [comment]
Thanks!
